core/store/src/store.rs

```diff
@@ -81,30 +81,43 @@ impl Store {
         {
             let mut lock = cache.lock();
             if let Some(value) = lock.values.get(key) {
-                match Arc::downcast::<T>(Arc::clone(value)) {
-                    Ok(result) => return Ok(Some(result)),
-                    Err(_) => {
-                        tracing::debug!(
-                            target: "store",
-                            requested = std::any::type_name::<T>(),
-                            "could not downcast an available cached deserialized value"
-                        );
+                if let Some(value) = value {
+                    // If the value is already cached, try to downcast it to the requested type.
+                    // If it fails, we log a debug message and continue to fetch from the database.
+                    match Arc::downcast::<T>(Arc::clone(value)) {
+                        Ok(result) => return Ok(Some(result)),
+                        Err(_) => {
+                            tracing::debug!(
+                                target: "store",
+                                requested = std::any::type_name::<T>(),
+                                "could not downcast an available cached deserialized value"
+                            );
+                        }
                     }
+                } else {
+                    // Value is cached as `None`, which means it was previously fetched
+                    // but was not found in the database.
+                    return Ok(None);
                 }
             }
         }
 
         let value = match self.get_ser::<T>(column, key) {
-            Ok(Some(value)) => Arc::from(value),
-            Ok(None) => return Ok(None),
+            Ok(Some(value)) => Some(Arc::from(value)),
+            Ok(None) => None,
             Err(e) => return Err(e),
         };
 
         let mut lock = cache.lock();
         if lock.active_flushes == 0 {
-            lock.values.put(key.into(), Arc::clone(&value) as _);
+            if let Some(v) = value.as_ref() {
+                lock.values.put(key.into(), Some(Arc::clone(v) as _));
+            } else if lock.store_none_values() {
+                // If the cache is configured to store `None` values, we store it.
+                lock.values.put(key.into(), None);
+            }
```

Comment:
I looked at putting `put` & `get` on the cache itself, but since the logic to alter `active_flushes` is in this file, it seems the types are still coupled (i.e.,`store` knows about the cache internals). In any case I am open to alternatives or suggestions here.